### DISCUSSION TITLE
How to Calculate the Partial Differential of 4 Momentum ？

### DISCUSSION THREAD

**Discussion by iisiaccNewton** [FeynCalc/feyncalc#260] [Q&A]
hello FeynCalc
Can we calculate the partial differential in the following figure ？
![屏幕截图 2024-04-10 194951](https://github.com/FeynCalc/feyncalc/assets/112673778/9c925ade-25bb-41fe-a716-e36515bf2c0d)

If this function does not exist, how to add some custom functions ?

**Comment by iisiaccNewton** [comment]
I found that the following code runs very well, but I don't know how to add it into FeynCalc. It is very troublesome to run this code before each calculation.
{
 pd[FV[p_, b_], p_, a_] := MT[a, b];
 pd[GS[p_], p_, a_] := GA[a];
 pd[SP[p_], p_, a_] := 2 FV[p, a];
 pd[SP[p_, k_], p_, a_] := FV[k, a];
 pd[f1_ + f2_, p_, a_] := pd[f1, p, a] + pd[f2, p, a];
 pd[f_^n_, p_, a_] := n*f^(n - 1)*pd[f, p, a];
 pd[f1_*f2_, p_, a_] := pd[f1, p, a]*f2 + f1*pd[f2, p, a];
 pd[Dot[f1_, f2_], p_, 
   a_] := (Dot[pd[f1, p, a], f2] + 
     Dot[f1, pd[f2, p, a]]) /. {Dot[___, 0, ___] :> 0, 
    Dot[s___, g_ + c_, h___] :> Dot[s, g, h] + Dot[s, c, h]}; 
 pd[f_, p_, a_] := If[FreeQ[f, p], 0, pd[f, p, a]];
 
 }

Here is an example.
![屏幕截图 2024-04-10 185351](https://github.com/FeynCalc/feyncalc/assets/112673778/fab4fdff-1940-4952-9c2f-32b442558482)

**Comment by vsht** [comment]
FourDivergence should be sufficient

https://feyncalc.github.io/FeynCalcBookDev/FourDivergence.html

Am 10.04.24 um 12:54 schrieb Isaac Newton:

**Reply by iisiaccNewton** [reply]
I find the function names you define very confusing, which often makes it difficult for me to find them when consulting manuals.
in this example,i think defining function names as PartialD is more convenient for searching
